### PR TITLE
Refactors cosmosdb provisioning to be triggered by Terraform

### PR DESCRIPTION
**Pull request by cloudify**
Example execution:

```
azurerm_resource_group.azurerm_resource_group: Refreshing state... (ID: /subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test)
azurerm_cosmosdb_account.azurerm_cosmosdb: Refreshing state... (ID: /subscriptions/ec285037-c673-4f58-b594-...DB/databaseAccounts/agid-cosmosdb-test)
null_resource.azurerm_cosmosdb_collections[1]: Refreshing state... (ID: 3110515839865585367)
null_resource.azurerm_cosmosdb_collections[3]: Refreshing state... (ID: 517883575258370935)
null_resource.azurerm_cosmosdb_collections[0]: Refreshing state... (ID: 8110838832867959983)
null_resource.azurerm_cosmosdb_collections[2]: Refreshing state... (ID: 6718070766871692562)

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

-/+ null_resource.azurerm_cosmosdb_collections[0] (new resource required)
      id:                           "8110838832867959983" => <computed> (forces new resource)
      triggers.%:                   "3" => "3"
      triggers.collections_json:    "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
      triggers.cosmosdb_id:         "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
      triggers.provisioner_version: "4" => "5" (forces new resource)

-/+ null_resource.azurerm_cosmosdb_collections[1] (new resource required)
      id:                           "3110515839865585367" => <computed> (forces new resource)
      triggers.%:                   "3" => "3"
      triggers.collections_json:    "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
      triggers.cosmosdb_id:         "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
      triggers.provisioner_version: "4" => "5" (forces new resource)

-/+ null_resource.azurerm_cosmosdb_collections[2] (new resource required)
      id:                           "6718070766871692562" => <computed> (forces new resource)
      triggers.%:                   "3" => "3"
      triggers.collections_json:    "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
      triggers.cosmosdb_id:         "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
      triggers.provisioner_version: "4" => "5" (forces new resource)

-/+ null_resource.azurerm_cosmosdb_collections[3] (new resource required)
      id:                           "517883575258370935" => <computed> (forces new resource)
      triggers.%:                   "3" => "3"
      triggers.collections_json:    "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
      triggers.cosmosdb_id:         "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
      triggers.provisioner_version: "4" => "5" (forces new resource)


Plan: 4 to add, 0 to change, 4 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.azurerm_cosmosdb_collections[3]: Destroying... (ID: 517883575258370935)
null_resource.azurerm_cosmosdb_collections[1]: Destroying... (ID: 3110515839865585367)
null_resource.azurerm_cosmosdb_collections[2]: Destroying... (ID: 6718070766871692562)
null_resource.azurerm_cosmosdb_collections[0]: Destroying... (ID: 8110838832867959983)
null_resource.azurerm_cosmosdb_collections[1]: Destruction complete after 0s
null_resource.azurerm_cosmosdb_collections[0]: Destruction complete after 0s
null_resource.azurerm_cosmosdb_collections[3]: Destruction complete after 0s
null_resource.azurerm_cosmosdb_collections[2]: Destruction complete after 0s
null_resource.azurerm_cosmosdb_collections[3]: Creating...
  triggers.%:                   "" => "3"
  triggers.collections_json:    "" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
  triggers.cosmosdb_id:         "" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
  triggers.provisioner_version: "" => "5"
null_resource.azurerm_cosmosdb_collections[1]: Creating...
  triggers.%:                   "" => "3"
  triggers.collections_json:    "" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
  triggers.cosmosdb_id:         "" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
  triggers.provisioner_version: "" => "5"
null_resource.azurerm_cosmosdb_collections[0]: Creating...
  triggers.%:                   "" => "3"
  triggers.collections_json:    "" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
  triggers.cosmosdb_id:         "" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
  triggers.provisioner_version: "" => "5"
null_resource.azurerm_cosmosdb_collections[1]: Provisioning with 'local-exec'...
null_resource.azurerm_cosmosdb_collections[0]: Provisioning with 'local-exec'...
null_resource.azurerm_cosmosdb_collections[3]: Provisioning with 'local-exec'...
null_resource.azurerm_cosmosdb_collections[2]: Creating...
  triggers.%:                   "" => "3"
  triggers.collections_json:    "" => "{\"messages\":\"fiscalCode\",\"notifications\":\"messageId\",\"profiles\":\"fiscalCode\",\"services\":\"serviceId\"}"
  triggers.cosmosdb_id:         "" => "/subscriptions/ec285037-c673-4f58-b594-d7c480da4e8b/resourceGroups/agid-rg-test/providers/Microsoft.DocumentDB/databaseAccounts/agid-cosmosdb-test"
  triggers.provisioner_version: "" => "5"
null_resource.azurerm_cosmosdb_collections[3] (local-exec): Executing: ["/bin/sh" "-c" "ts-node infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts -g agid-rg-test -n agid-cosmosdb-test -d agid-documentdb-test -c services -k serviceId"]
null_resource.azurerm_cosmosdb_collections[2]: Provisioning with 'local-exec'...
null_resource.azurerm_cosmosdb_collections[0] (local-exec): Executing: ["/bin/sh" "-c" "ts-node infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts -g agid-rg-test -n agid-cosmosdb-test -d agid-documentdb-test -c messages -k fiscalCode"]
null_resource.azurerm_cosmosdb_collections[1] (local-exec): Executing: ["/bin/sh" "-c" "ts-node infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts -g agid-rg-test -n agid-cosmosdb-test -d agid-documentdb-test -c notifications -k messageId"]
null_resource.azurerm_cosmosdb_collections[2] (local-exec): Executing: ["/bin/sh" "-c" "ts-node infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts -g agid-rg-test -n agid-cosmosdb-test -d agid-documentdb-test -c profiles -k fiscalCode"]
null_resource.azurerm_cosmosdb_collections[1] (local-exec): info: Making sure database exists: name=agid-documentdb-test
null_resource.azurerm_cosmosdb_collections[0] (local-exec): info: Making sure database exists: name=agid-documentdb-test
null_resource.azurerm_cosmosdb_collections[3] (local-exec): info: Making sure database exists: name=agid-documentdb-test
null_resource.azurerm_cosmosdb_collections[2] (local-exec): info: Making sure database exists: name=agid-documentdb-test
null_resource.azurerm_cosmosdb_collections[0] (local-exec): info: Making sure collection exists: name=messages partitionKey=fiscalCode
null_resource.azurerm_cosmosdb_collections[0] (local-exec): info: Create CosmosDB collection messages with partitionKey fiscalCode
null_resource.azurerm_cosmosdb_collections[0] (local-exec): info: Completed
null_resource.azurerm_cosmosdb_collections[0]: Creation complete after 5s (ID: 7063553211980026950)
null_resource.azurerm_cosmosdb_collections[3] (local-exec): info: Making sure collection exists: name=services partitionKey=serviceId
null_resource.azurerm_cosmosdb_collections[3] (local-exec): info: Create CosmosDB collection services with partitionKey serviceId
null_resource.azurerm_cosmosdb_collections[1] (local-exec): info: Making sure collection exists: name=notifications partitionKey=messageId
null_resource.azurerm_cosmosdb_collections[1] (local-exec): info: Create CosmosDB collection notifications with partitionKey messageId
null_resource.azurerm_cosmosdb_collections[3] (local-exec): info: Completed
null_resource.azurerm_cosmosdb_collections[1] (local-exec): info: Completed
null_resource.azurerm_cosmosdb_collections[3]: Creation complete after 5s (ID: 4159500268542967754)
null_resource.azurerm_cosmosdb_collections[1]: Creation complete after 5s (ID: 266564979161739920)
null_resource.azurerm_cosmosdb_collections[2] (local-exec): info: Making sure collection exists: name=profiles partitionKey=fiscalCode
null_resource.azurerm_cosmosdb_collections[2] (local-exec): info: Create CosmosDB collection profiles with partitionKey fiscalCode
null_resource.azurerm_cosmosdb_collections[2] (local-exec): info: Completed
null_resource.azurerm_cosmosdb_collections[2]: Creation complete after 5s (ID: 1763508215655198889)

Apply complete! Resources: 4 added, 0 changed, 4 destroyed.
```

### CODE REVIEW

**Comment by cloudify** [comment]
Since collections have their own resource, we can tell Terraform to apply changes only to that resource, much control!

``` shell
terraform apply -var-file=infrastructure/env/common/tfvars.json -var-file=infrastructure/env/test/tfvars.json -target=null_resource.azurerm_cosmosdb_collections infrastructure
```

File: infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts [hunk 1]
```diff
@@ -112,53 +101,89 @@ const createCollectionIfNotExists = (
...
-  )
-  .catch((e: Error) => console.error(process.env.VERBOSE ? e : e.message));
+// check whether all required environment variables are set
+const missingEnvs = missingLoginEnvironment();
+if (missingEnvs.length > 0) {
+  console.error(`Missing required env vars: ${missingEnvs.join(", ")}`);
+  process.exit(-1);
+}
+
+const argv = yargs
```
Comment (gunzip) [inline infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts]: consider also commander which has typings for TS:

https://stackoverflow.com/questions/37199787/how-can-i-use-commander-npm-with-typescript

File: infrastructure/azure.tf [hunk 2]
```diff
@@ -244,6 +262,29 @@ resource "azurerm_cosmosdb_account" "azurerm_cosmosdb" {
...
+    # increment the following value when changing the provisioner script to
+    # trigger the re-execution of the script
+    # TODO: consider using the hash of the script content instead
+    provisioner_version = "5"
+  }
+
+  count = "${length(keys(var.azurerm_cosmosdb_collections))}"
+
+  provisioner "local-exec" {
+    command = "ts-node ${var.cosmosdb_collection_provisioner} -g ${azurerm_resource_group.azurerm_resource_group.name} -n ${azurerm_cosmosdb_account.azurerm_cosmosdb.name} -d ${var.azurerm_cosmosdb_documentdb} -c ${element(keys(var.azurerm_cosmosdb_collections), count.index)} -k ${lookup(var.azurerm_cosmosdb_collections, element(keys(var.azurerm_cosmosdb_collections), count.index))}"
```
Comment (gunzip) [inline infrastructure/azure.tf]: better use the longopts ?

File: .node-version [hunk 3]
```diff
@@ -1 +1 @@
-6.5.0
+8.5.0
```
Comment (gunzip) [inline .node-version]: why ? (I've node 6.5.x installed locally to mimic the Azure Function environment)

File: infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts (hunk 1 above)
Comment (cloudify) [reply infrastructure/local-provisioners/azurerm_cosmosdb_collection.ts]: what do you mean? it looks like `yargs`

File: .node-version (hunk 3 above)
Comment (cloudify) [reply .node-version]: ops removed

File: infrastructure/azure.tf (hunk 2 above)
Comment (cloudify) [reply infrastructure/azure.tf]: done

**Pull request merged**